### ISSUE TITLE
Cannot find a template function definition in rapidxml

### ISSUE THREAD

**Issue by junwha0511** [github/codeql#17569]
Hi,

```
#include "rapidxml.hpp"
#include "rapidxml_utils.hpp"

int main() {
 rapidxml::file<> xmlFile("example.xml"); 
}
```

For this simple example, I observe that all functions including `allocate_aligned` are selectable with the simple Function query,
```
from Function f
select f, f.getName()
```

But definition is not available for the most of functions in rapidxml.hpp, and also for `allocate_aligned`.
https://sourcegraph.com/github.com/dwd/rapidxml@v1.0.0/-/blob/rapidxml.hpp?L677

```
from Function f
where f.hasDefinition() 
select f, f.getName()
```

But it is obvious that all the functions in rapidxml.hpp is compiled, not linked with declarations-only.

I think it occurs for member functions in a template class.
some template member functions are incorrectly identified as having a no-definition, though it has a definition exactly in the result of the first query.

Could you help me know what is the problem of this?

**Comment by smowton** [comment]
This is as expected -- template functions, and function members of template classes, are elaborated on demand. If you want to inspect a function instance that isn't naturally reachable from the program you're extracting, you might consider extracting a dummy source file that deliberately forces elaboration of the desired method. For example, allocate_aligned is private but is used from the public allocate_node method; thus I can use

```
#include "rapidxml.hpp"

rapidxml::memory_pool<> pool;

void test() {
  pool.allocate_node(rapidxml::node_type::node_document);
}
```

Then after `codeql database create -l cpp -c "g++ test.cpp -c -o test.o" db` I can see `allocate_aligned` now shows up as a definition with a body as expected.

**Comment by smowton** [comment]
A followup with a bit more detail:

There are a few different forks / versions of this library, and the results you'll see vary from version to version:

* https://github.com/Fe-Bell/RapidXML
* https://github.com/dwd/rapidxml/releases/tag/v1.0.0 (which your sourcegraph link above points to)
* https://github.com/dwd/rapidxml/tree/master

In the first two, the function `allocate_aligned` is an ordinary (non-templated) function inside a templated class. This behaves as I described above: it isn't present in the database unless demanded by some elaboration / instantiation of its containing template class, as in my `pool.allocate_node` call.

In the last one, the function `allocate_aligned` is a templated function (in addition to still sitting inside a templated class). This **will** generated an uninstantiated version in the database, even if user code doesn't touch it (e.g. the input user program can just be `#include <rapidxml.hpp>`).

This distinction between the two scenarios (ordinary-function-in-templated-class vs templated-function-in-templated-class) may or may not constitute a bug; we'll look into it in a bit more detail. Meanwhile the suggested workaround, to nudge template instantiation with an appropriate call in some dummy code, continues to apply.

**Comment by junwha0511** [comment]
Hi, thank you for the detailed comments!

You mean **if the function is not used, it would be a Function class in DB that does not have a definition although it actually is**, right? (here, I think it is weird that Function is visible but definition is not visible.)

More specifically, I'm analyzing cocos2d-x(https://github.com/cocos2d/cocos2d-x) repository, but it does not contain the rapidxml::memory_pool::allocate_aligned definition though it seems using this function.

At [SAXParser::parseIntrusive](https://sourcegraph.com/github.com/cocos2d/cocos2d-x/-/blob/cocos/platform/CCSAXParser.cpp?L107)

it creates rapidxml::xml_sax3_parser<> parser(&printer) and parses using parser.parse<>(),
which is used by `SAXParser::parse`, and this `SAXParser::parse` is used in many places (https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/cocos2d/cocos2d-x%24+parser.parse&patternType=keyword&sm=0), thus it is obvious that rapidxml parsers are used, and reachable.

I confirmed that that allocate_aligned is invoked by rapidxml::xml_sax3_parser<> parser(&printer) and parser.parse<>() by simple example (mocking SAXParser in cocos2d-x) and also can be found the definition of allocate_aligned by CodeQL. 
```
#include<iostream>
#include "rapidxml_sax3.hpp"
#include<stdlib.h>
#include <fstream>
#include <sstream>

class RapidXmlSaxHander : public rapidxml::xml_sax2_handler
{

public:
    RapidXmlSaxHander() {};

    void xmlSAX2StartElement(const char *name, size_t /*len*/, const char **atts, size_t /*attslen*/) override {}

    void xmlSAX2EndElement(const char *name, size_t /*len*/) override {}

    void xmlSAX2Text(const char *s, size_t len) override {}
};

int main() {
    // Load the XML file into memory
    std::ifstream file("example.xml");
    if (!file) {
        std::cerr << "Failed to open file." << std::endl;
        return 1;
    }

    std::stringstream buffer;
    buffer << file.rdbuf(); // Read the file content into buffer
    std::string xmlData = buffer.str();
    
    RapidXmlSaxHander printer;
    rapidxml::xml_sax3_parser<> parser(&printer);
    parser.parse<0>(xmlData.data(), static_cast<int>(xmlData.size()));

}
```

But the almost same code, and used code, in cocos2d-x, it has no definition, but Function is visible. What would be wrong?

**Comment by junwha0511** [comment]
FYI: here's the command I built cocos2d-x
<details>
<summary>Build Command</summary>

```
#!/bin/bash
export CC=/usr/bin/clang CXX=/usr/bin/clang++
git clone https://github.com/cocos2d/cocos2d-x
cd cocos2d-x
git checkout 76903dee64046c7bfdba50790be283484b4be271

sed -i 's#git://#https://#' .gitmodules
echo -e "y\n" | ./install-deps-linux.sh
echo -e "no\n" | python3 download-deps.py
git submodule update --init

# conflict avoidance for Ubuntu 22.04 (https://github.com/cocos2d/cocos2d-x/issues/20471)
wget https://github.com/cocos2d/cocos2d-x/files/4272227/libchipmunk7.0.1.zip
rm -rf external/chipmunk/prebuilt/linux/64-bit
unzip libchipmunk7.0.1.zip -d external/chipmunk/prebuilt/linux/64-bit

cmake -B build -S .
cmake --build build

# codeql database create cocos2d-x_db --language=c-cpp --command="./build.sh" 
```
</details>

**Comment by smowton** [comment]
Reading https://github.com/cocos2d/cocos2d-x-3rd-party-libs-bin/blob/v3/rapidxml/rapidxml_sax3.hpp, which appears to be the version of rapidxml_sax3.hpp that cocos2d is using, I am not convinced that the sax3 implementation there actually uses `rapidxml::memory_pool`. For example we see misleading comments like

```
    //! This class represents root of the DOM hierarchy. 
    //! It is also an xml_node and a memory_pool through public inheritance.
    ...
    template<class Ch = char>
    class xml_sax3_parser
```

But this comment is clearly copy-pasted from the documentation for `class xml_document : public xml_node<Ch>, public memory_pool<Ch>` in the "real" `rapidxml.hpp`, and in fact the sax3 code does not inherit from `memory_pool`.

Similarly in places where we'd expect to see an `allocate_node` -> `allocate_aligned` call, like in `parse_element`, in the sax3 parser we see:

```
        // Parse element node
        template<int Flags>
        void parse_element(Ch *&text)
        {
            // Create element node
            // xml_node<Ch> *element = this->allocate_node(node_element);
```

The allocation code has been commented out, which makes sense considering the SaX family of APIs are based around making callbacks at particular points in parsing an XML element, *not* in constructing a DOM tree, which would require a node allocator.

Basically it looks as if the rapidxml sax3 implementation has been created by hurried hacking at the original rapidxml DOM parser, leaving misleading unaltered comments from the DOM version, commented-out code that used to perform allocation, `#if 0` blocks in the same spirit, and so on.

So, I think it's probably right to conclude that if cocos2d is only using the sax3 parser, it likely is not touching `rapidxml::memory_pool` in any real sense.

**Comment by junwha0511** [comment]
I see, I missed "#if 0 // disable data node" above allocate_node...
As you said, It looks like cocos2d-x disabled the memory pool in their SAX-customized rapidxml. 

I apologize for the confusion and appreciate your support.
